lib/cancan/rule.rb

```diff
@@ -73,12 +76,31 @@ def matches_subject?(subject)
       @subjects.include?(:all) || @subjects.include?(subject) || matches_subject_class?(subject)
     end
 
+    def matches_attribute?(attribute)
+      return true if @attributes.empty?
+      return @base_behavior if attribute.nil?
+      @attributes.include?(attribute.to_sym)
+    end
+
     def matches_subject_class?(subject)
       @subjects.any? do |sub|
         sub.is_a?(Module) && (subject.is_a?(sub) ||
           subject.class.to_s == sub.to_s ||
           (subject.is_a?(Module) && subject.ancestors.include?(sub)))
       end
     end
+
+    def parse_attributes_from_extra_args(args)
+      attributes = args.shift if valid_attribute_param?(args.first)
+      extra_args = args.shift
+
+      [attributes, extra_args]
+    end
+
+    def condition_and_block_check(conditions, block, action, subject)
+      return unless conditions.is_a?(Hash) && block
+      raise BlockAndConditionsError, 'A hash of conditions is mutually exclusive with a block.'\
```

Comment:
this part could be extracted into a separate PR and immediately merged to make this PR smaller